### PR TITLE
Feature: configurable custom file exports

### PR DESCRIPTION
**Pull request by KarolakJakub**
This PR includes a feature that allows to create custom export options, as described in configuration.md - users will be able to add multiple configurable export types with different formatting, locales etc. If no defined exports are found within the config, Turnilo will use the default 'Export to TSV' and 'Export to CSV' ones.

### CODE REVIEW

File: src/common/models/customization/customization.ts [hunk 1]
```diff
@@ -100,6 +104,9 @@ export class Customization implements Instance<CustomizationValue, Customization
       value.urlShortener = UrlShortener.fromJS(parameters.urlShortener);
     }
 
+    const options = new ClassShareOptions(parameters.shareOptions);
+    value.shareOptions = options.getOptions();
```
Comment (adrianmroz) [inline src/common/models/customization/customization.ts]: Looks suspicious. Why then separate `ShareOptions` type and class?

File: src/common/models/app-settings/app-settings.ts [hunk 2]
```diff
@@ -23,6 +23,7 @@ import { Cluster, ClusterJS } from "../cluster/cluster";
 import { Customization, CustomizationJS } from "../customization/customization";
 import { DataCube, DataCubeJS } from "../data-cube/data-cube";
 import { Manifest } from "../manifest/manifest";
+import { shareOptionsDefaults } from "../share-options/share-options";
```
Comment (adrianmroz) [inline src/common/models/app-settings/app-settings.ts]: Unused import?

File: src/client/utils/tabular-options/tabular-options.ts [hunk 3]
```diff
@@ -34,15 +37,13 @@ function findSeriesAndDerivation(name: string, concreteSeriesList: List<Concrete
   return null;
 }
 
-export default function tabularOptions(essence: Essence): TabulatorOptions {
+export default function tabularOptions(essence: Essence, locales?: Locales): TabulatorOptions {
```
Comment (adrianmroz) [inline src/client/utils/tabular-options/tabular-options.ts]: You use `locales` like it is required.

File: src/client/utils/download/download.ts [hunk 4]
```diff
@@ -37,19 +37,20 @@ export function getMIMEType(fileType: string) {
...
   const type = `${getMIMEType(fileFormat)};charset=utf-8`;
   const blob = new Blob([datasetToFileString(dataset, fileFormat, options)], { type });
   if (!fileName) fileName = `${new Date()}-data`;
   fileName += `.${fileFormat}`;
   filesaver.saveAs(blob, fileName, true); // true == disable auto BOM
 }
 
-export function datasetToFileString(dataset: Dataset, fileFormat: FileFormat, options?: TabulatorOptions): string {
+export function datasetToFileString(dataset: Dataset, fileFormat?: FileFormat, options?: TabulatorOptions): string {
+  const optionsCopy = { ...options };
```
Comment (adrianmroz) [inline src/client/utils/download/download.ts]: Possible spread on undefined.

File: src/client/utils/download/download.ts [hunk 5]
```diff
@@ -37,19 +37,20 @@ export function getMIMEType(fileType: string) {
   }
 }
 
-export function download({ dataset, options }: DataSetWithTabOptions, fileFormat: FileFormat, fileName?: string): void {
+export function download({ dataset, options }: DataSetWithTabOptions, fileFormat: FileFormat, fileName?: string, custom?: boolean): void {
```
Comment (adrianmroz) [inline src/client/utils/download/download.ts]: Unused param

File: src/client/components/share-menu/share-menu.tsx [hunk 6]
```diff
@@ -116,8 +131,8 @@ function externalViewItems({ customization: { externalViews = [] }, essence }: E
 }
 
 export const ShareMenu: React.SFC<ShareMenuProps> = props => {
-  const { openOn, onClose } = props;
-
+  const { openOn, onClose, customization } = props;
+  const shareOptions = customization.shareOptions;
```
Comment (adrianmroz) [inline src/client/components/share-menu/share-menu.tsx]: Unused variable.

File: src/client/components/share-menu/share-menu.tsx [hunk 7]
```diff
@@ -40,25 +42,38 @@ export interface ShareMenuProps {
...
 
-function onExport(fileFormat: FileFormat, props: ExportProps) {
+function onExport(shareOption: ShareOption, props: ExportProps) {
+
+  const { separator, lineBreak, finalLineBreak, columnOrdering, format, locale } = shareOption;
   const { onClose, getDownloadableDataset, essence, timekeeper } = props;
-  const dataSetWithTabOptions = getDownloadableDataset();
+
+  let dataSetWithTabOptions = getDownloadableDataset();
+  dataSetWithTabOptions.options = {
```
Comment (adrianmroz) [inline src/client/components/share-menu/share-menu.tsx]: Extract function that deals with construction of this object.

**Comment by adrianmroz** [comment]
Let's extract bug fix to another PR, so we could merge it earlier. Rest of it should wait till we discuss what is needed and how we should achieve it. For reference #504

**Comment by KarolakJakub** [comment]
Bugfix extracted to https://github.com/allegro/turnilo/pull/536

File: docs/configuration.md [hunk 8]
```diff
@@ -699,6 +699,48 @@ Advanced options example:
...
 You can define a `customization:` section in the config to configure some aspects of the look and feel of Turnilo.
 
+You can define custom share options section (`shareOptions:`), which will force specific formatting on exported files. It will ovewrite two default CSV and TSV exports in the share menu. You can define specific properties for number, time and time range types in the `locale:` section. Locale and Locale Options sections are optional but they will be set to default values if not provided.
+
+```yaml
+  shareOptions:
+
+  - title: "excel friendly export to TSV" # name displayed in the share menu
+    format: # "tsv" | "csv"
+    separator: # "\t" | "," | ...
```
Comment (mkuthan) [inline docs/configuration.md]: What is the reason to define format AND separator? Format defines separator implicitly.

File: docs/configuration.md [hunk 9]
```diff
@@ -699,6 +699,48 @@ Advanced options example:
...
+
+```yaml
+  shareOptions:
+
+  - title: "excel friendly export to TSV" # name displayed in the share menu
+    format: # "tsv" | "csv"
+    separator: # "\t" | "," | ... 
+    lineBreak: "\r\n"
+    finalLineBreak: # "include" | "suppress"
+    columnOrdering: # "keys-first" | "as-seen";
```
Comment (mkuthan) [inline docs/configuration.md]: Export should always preserve what is presented on the screen. If not user will be confused, please remember that end user is not aware of configuration.

File: docs/configuration.md [hunk 10]
```diff
@@ -699,6 +699,48 @@ Advanced options example:
...
+    locale:
```
Comment (mkuthan) [inline docs/configuration.md]: As Adrian mentioned, there is open issue #504. And again, export should always preserve what is presented on the screen. It applies also to number and date formats. If some formats are not currently available on UI we should focus to  remove existing limitations and add more formatting options. 

Turnilo should be easy to understand for end user. How to

Comment (KarolakJakub) [reply docs/configuration.md]: I've replied in the #504 . :+1: 
Thank you for review, I will apply and push requested changes soon.
Regarding exports reflecting what's presented on screen/app - please note that this is already not the case. Turnilo is modifying date format on export https://github.com/allegro/turnilo/blob/8824d86d37e354baea45cff492e7e57154daab5d/src/client/utils/tabular-options/tabular-options.ts#L40 so we do get different values.

Consider this export made on latest master - data is in a different format and value is rounded up in the app view.
image

Comma used as decimal separator is our main business issue - Turnilo uses dot to separate fractional part of a number which causes issues when opened in Excel in our locale (pl-PL).

Comment (mkuthan) [reply docs/configuration.md]: Yep, date & number formats are fixed in CSV/TSV. Currently there is no way to configure format for date dimensions in Turnilo, so dates on UI are formatted in the most concise way. We also put some effort (#288) to keep date formatting consistent across whole app, the export is the only exception with ISO formatting.

On the other hand for metrics, administrator is able to configure default format and user is able to change the configured format on UI, but the formatting is totally ignored by export CSV/TSV :(

Right now I don't know how to design date/number formatting in consistent way. I feel convinced that formatting on UI and formatting in the exports could be different but the formatting is rather related to the given dimension/measure not to particular export. 

Perhaps we should enhance existing formatting configuration to datetime and numeric dimensions in similar way to measures. But for sure start with good design before implementation step.

**Pull request closed without merging**